### PR TITLE
Add note for mismatched types because of circular dependencies

### PR DESCRIPTION
**Pull request by lukas-code**
If you have crate A with a dependency on crate B, and crate B with a dev-dependency on A, then you might see "mismatched types" errors on types that seem to be equal. This PR adds a note that explains that the types are different, because crate B is compiled twice, one time with `cfg(test)` and one time without.

I haven't found a good way to create circular dependencies in UI tests, so I abused the incremental tests instead. As a bonus, incremental tests support "cpass" now.

related to https://github.com/rust-lang/rust/issues/22750

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -379,7 +381,7 @@ impl<'test> TestCx<'test> {
         }
     }
 
-    fn run_rpass_test(&self) {
+    fn run_pass_test(&self, run: bool) {
         let emit_metadata = self.should_emit_metadata(self.pass_mode());
         let should_run = self.run_if_enabled();
         let proc_res = self.compile_test(should_run, emit_metadata);
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs:387]: This is passing misleading input into compile_test via should_run, I'm not sure if we should try to reuse run_pass_test across run and not-run modes like this. At minimum coupling `run` into `should_run` probably is necessary.

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -379,6 +381,20 @@ impl<'test> TestCx<'test> {
...
+        let emit_metadata = self.should_emit_metadata(self.pass_mode());
+        let proc_res = self.compile_test(WillExecute::No, emit_metadata);
+
+        if !proc_res.status.success() {
+            self.fatal_proc_rec("compilation failed!", &proc_res);
+        }
+
+        // FIXME(#41968): Move this check to tidy?
+        if !errors::load_errors(&self.testpaths.file, self.revision).is_empty() {
+            self.fatal("run-pass tests with expected warnings should be moved to ui/");
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs:394]:
```suggestion
            self.fatal("compile-pass tests with expected warnings should be moved to ui/");
```

**Comment by Mark-Simulacrum** [comment]
Apologies for the delay in reviewing @lukas-code, unless labels are updated I at least don't see PRs.

@rustbot author

**Comment by lukas-code** [comment]
I rebased and addressed the review comments.

@rustbot ready

**Pull request merged**